### PR TITLE
dx12: remap vertex bindings

### PR DESCRIPTION
**Pull request by msiglreith**
Map input attribute id to input binding to work around issues where attribute offsets are larger than the vertex stride. Current mapping strategy is naive, mapping one attribute to one binding. Better packing would be useful but only provides performance improvements.

Tested against several samples and the CTS texture reftests look reasonable but fail due to lack of freeDescriptorSets implementation.

PR checklist:
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: dx12

### CODE REVIEW

File: src/backend/dx12/src/command.rs [hunk 1]
```diff
@@ -591,25 +591,36 @@ impl CommandBuffer {
...
                     Some(start_offset) => {
                         let start_slot = last_end_slot + start_offset;
-                        let num_views = vbs_needs_bind[start_slot..]
+                        let num_views = vbs_remap[start_slot..]
                             .iter()
-                            .position(|needs_bind| !*needs_bind)
-                            .unwrap_or(vbs_needs_bind.len());
+                            .position(|remap| remap.is_none())
+                            .unwrap_or(vbs_remap.len());
+                        let buffers = (start_slot..start_slot+num_views).map(|i| {
```
Comment (kvark) [inline src/backend/dx12/src/command.rs]: nit: iterate slice of `vbs_remap[start_slot .. ]` and break on `None` instad?

File: src/backend/dx12/src/command.rs [hunk 2]
```diff
@@ -591,25 +591,36 @@ impl CommandBuffer {
...
+                            let remapping = vbs_remap[i].unwrap();
+                            let view = vbs[remapping.mapped_binding];
+
+                            d3d12::D3D12_VERTEX_BUFFER_VIEW {
+                                BufferLocation: view.BufferLocation + remapping.offset as u64,
+                                SizeInBytes: view.SizeInBytes - remapping.offset,
+                                StrideInBytes: remapping.stride,
+                            }
+                        }).collect::<Vec<_>>();
```
Comment (kvark) [inline src/backend/dx12/src/command.rs]: could we re-use the vector allocation at least within this function?

File: src/backend/dx12/src/command.rs [hunk 3]
```diff
@@ -591,25 +591,36 @@ impl CommandBuffer {
...
+                            .position(|remap| remap.is_none())
+                            .unwrap_or(vbs_remap.len());
+                        let buffers = (start_slot..start_slot+num_views).map(|i| {
+                            let remapping = vbs_remap[i].unwrap();
+                            let view = vbs[remapping.mapped_binding];
+
+                            d3d12::D3D12_VERTEX_BUFFER_VIEW {
+                                BufferLocation: view.BufferLocation + remapping.offset as u64,
+                                SizeInBytes: view.SizeInBytes - remapping.offset,
+                                StrideInBytes: remapping.stride,
```
Comment (kvark) [inline src/backend/dx12/src/command.rs]: doesn't it cause d3d12 validation warnings, given that offset + N * stride is outside the view range?

File: src/backend/dx12/src/device.rs [hunk 4]
```diff
@@ -1505,8 +1507,8 @@ impl d::Device<B> for Device {
                             return Some(Err(pso::CreationError::Other));
                         }
                     },
-                    InputSlot: attrib.binding as _,
-                    AlignedByteOffset: attrib.element.offset,
+                    InputSlot: i as _,
+                    AlignedByteOffset: 0, // Rebased during the vertex binding remapping
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: Wait, so you are force remapping everything to have zero attribute offset? That is quite rough and may cause us to run out of the vertex buffer could.

Comment (msiglreith) [reply src/backend/dx12/src/device.rs]: Never happened so far

File: src/backend/dx12/src/command.rs (hunk 3 above)
Comment (msiglreith) [reply src/backend/dx12/src/command.rs]: nop

File: src/backend/dx12/src/command.rs (hunk 1 above)
Comment (msiglreith) [reply src/backend/dx12/src/command.rs]: done but involved quite a bit of iterator origami

File: src/backend/dx12/src/command.rs (hunk 2 above)
Comment (msiglreith) [reply src/backend/dx12/src/command.rs]: Done via SmallVec avoiding allocations (I hope)

File: src/backend/dx12/src/command.rs [hunk 5]
```diff
@@ -601,25 +601,26 @@ impl CommandBuffer {
...
+                            .map(|mapping| {
+                                let view = vbs[mapping.mapped_binding];
+
+                                d3d12::D3D12_VERTEX_BUFFER_VIEW {
+                                    BufferLocation: view.BufferLocation + mapping.offset as u64,
+                                    SizeInBytes: view.SizeInBytes - mapping.offset,
+                                    StrideInBytes: mapping.stride,
+                                }
+                            })
+                            .collect::<SmallVec<[d3d12::D3D12_VERTEX_BUFFER_VIEW; MAX_VERTEX_BUFFERS]>>();
```
Comment (kvark) [inline src/backend/dx12/src/command.rs]: I don't think you need that actual type in there, could just be `SmallVec<_>`

File: src/backend/dx12/src/command.rs [hunk 6]
```diff
@@ -601,25 +601,26 @@ impl CommandBuffer {
...
-                            let remapping = vbs_remap[i].unwrap();
-                            let view = vbs[remapping.mapped_binding];
-
-                            d3d12::D3D12_VERTEX_BUFFER_VIEW {
-                                BufferLocation: view.BufferLocation + remapping.offset as u64,
-                                SizeInBytes: view.SizeInBytes - remapping.offset,
-                                StrideInBytes: remapping.stride,
-                            }
-                        }).collect::<Vec<_>>();
+                            .take_while(|x| x.is_some())
```
Comment (kvark) [inline src/backend/dx12/src/command.rs]: nice iterator usage :+1:

File: src/backend/dx12/src/device.rs [hunk 7]
```diff
@@ -1497,6 +1515,30 @@ impl d::Device<B> for Device {
                 };
                 let format = attrib.element.format;
 
+                // Check if we need to add a new remapping in-case the offset is
+                // higher than the vertex stride.
+                // In this case we rebase the attribute to zero offset.
+                let binding = attrib.binding as usize;
+                let stride = vertex_strides[binding];
+                let offset = attrib.element.offset;
+                let (input_slot, offset) = if stride <= offset {
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: how does it handle a case when `offset < stride` but `offset + element_size > stride`?

Comment (msiglreith) [reply src/backend/dx12/src/device.rs]: Works fine so far, added warden test for this.

File: src/backend/dx12/src/command.rs (hunk 5 above)
Comment (msiglreith) [reply src/backend/dx12/src/command.rs]: I could remove the array type at least but not the size

**Pull request merged**